src/http/SOCKSProxy.zig

```diff
@@ -0,0 +1,240 @@
+};
+
+pub fn create(allocator: std.mem.Allocator, proxy_url: URL, destination_host: []const u8, destination_port: u16) !*SOCKSProxy {
+    const socks_proxy = bun.new(SOCKSProxy, .{
+        .ref_count = .init(),
+        .proxy_url = proxy_url,
+        .destination_host = destination_host,
+        .destination_port = destination_port,
+        .allocator = allocator,
+    });
+
+    return socks_proxy;
+}
+
+pub fn sendAuthHandshake(this: *SOCKSProxy, comptime is_ssl: bool, socket: NewHTTPContext(is_ssl).HTTPSocket) void {
+    // SOCKS5 authentication handshake
+    // +----+----------+----------+
+    // |VER | NMETHODS | METHODS  |
+    // +----+----------+----------+
+    // | 1  |    1     | 1 to 255 |
+    // +----+----------+----------+
+    var auth_request = [_]u8{ @intFromEnum(SOCKSVersion.v5), 1, @intFromEnum(SOCKSAuthMethod.no_auth) };
+
+    _ = socket.write(&auth_request);
+    this.state = .auth_handshake;
+}
+
+pub fn sendConnectRequest(this: *SOCKSProxy, comptime is_ssl: bool, socket: NewHTTPContext(is_ssl).HTTPSocket) !void {
+    // SOCKS5 connect request
+    // +----+-----+-------+------+----------+----------+
+    // |VER | CMD |  RSV  | ATYP | DST.ADDR | DST.PORT |
+    // +----+-----+-------+------+----------+----------+
+    // | 1  |  1  | X'00' |  1   | Variable |    2     |
+    // +----+-----+-------+------+----------+----------+
+
+    var buffer = std.ArrayList(u8).init(this.allocator);
+    defer buffer.deinit();
+
+    // Version, Command, Reserved
+    try buffer.appendSlice(&[_]u8{ @intFromEnum(SOCKSVersion.v5), @intFromEnum(SOCKSCommand.connect), 0x00 });
+
+    // Address type and address
+    if (strings.isIPAddress(this.destination_host)) {
+        if (strings.indexOf(this.destination_host, ":")) |_| {
+            // IPv6
+            try buffer.append(@intFromEnum(SOCKSAddressType.ipv6));
+            const parsed = std.net.Ip6Address.parse(this.destination_host, 0) catch {
+                return error.InvalidIPv6Address;
+            };
+            try buffer.appendSlice(std.mem.asBytes(&parsed.sa.addr));
+        } else {
+            // IPv4
+            try buffer.append(@intFromEnum(SOCKSAddressType.ipv4));
+            const parsed = std.net.Ip4Address.parse(this.destination_host, 0) catch {
+                return error.InvalidIPv4Address;
+            };
+            try buffer.appendSlice(std.mem.asBytes(&parsed.sa.addr));
+        }
+    } else {
+        // Domain name
+        try buffer.append(@intFromEnum(SOCKSAddressType.domain_name));
+        if (this.destination_host.len > 255) {
+            return error.DomainNameTooLong;
+        }
+        try buffer.append(@intCast(this.destination_host.len));
+        try buffer.appendSlice(this.destination_host);
+    }
+
+    // Port (big-endian)
+    const port_bytes = std.mem.toBytes(std.mem.nativeToBig(u16, this.destination_port));
+    try buffer.appendSlice(&port_bytes);
+
+    // Send the request
+    _ = socket.write(buffer.items);
```

Comment:
This is definitely wrong and also slow. We need to buffer the data incase the write is incomplete.